### PR TITLE
Sema: Apply @escaping to typealiases with underlying function type

### PR DESCRIPTION
**Pull request by slavapestov**
Work in progress. @milseman all that remains is a diagnostic for removing redundant @escaping in non-parameter position, since we decided that's not allowed.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

**Comment by slavapestov** [comment]
@swift-ci Please smoke test os x

**Comment by gparker42** [comment]
This segfaults on the Linux Swift build and breaks the Linux libdispatch build. I'm going to revert in #4369 unless you have a fix in the next few minutes.

**Comment by slavapestov** [comment]
Fix is going through CI now.

**Comment by gparker42** [comment]
I assume you mean #4368. Does that fix the NSURLSession segfault, the Queue.swift compilation error, or both?

**Comment by slavapestov** [comment]
The Queue.swift compilation error is a separate issue. libdispatch needs an update. Working on that.

File: lib/Sema/TypeCheckType.cpp [hunk 1]
```diff
@@ -1402,6 +1402,38 @@ static bool diagnoseAvailability(Type ty, IdentTypeRepr *IdType, SourceLoc Loc,
   return false;
 }
 
+/// Whether the given DC is a noescape-by-default context, i.e. not a property
+/// setter
+static bool isDefaultNoEscapeContext(const DeclContext *DC) {
+  auto funcDecl = dyn_cast<FuncDecl>(DC);
+  return !funcDecl || !funcDecl->isSetter();
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckType.cpp]: Hm. What about index parameters of subscripts?

**Pull request merged**